ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,118 @@
+(*
+ * Copyright (C) 2006-2009 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Helpers
+
+module D = Debug.Make (struct let name = "tar_ext" end)
+
+open D
+
+let file_perm = 0o644
+
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let ( + ) = Int64.add
+
+exception Illegal_file_path of string
+
+exception Unsupported_file_type of string
+
+exception File_exceeds_max_size_limit of string
+
+exception File_incomplete
+
+exception File_parsed_failed
+
+let unpack dir acc_size max_size ifd head =
+  let assert_file_path_is_legal file_path =
+    if
+      List.exists
+        (fun s -> String.equal "." s || String.equal ".." s)
+        (String.split_on_char '/' file_path)
+    then
+      raise (Illegal_file_path file_path)
+  in
+  let assert_file_not_exceed_max_size acc_size =
+    debug "Current total unpacked file size: %Ld" acc_size ;
+    if Int64.compare acc_size max_size > 0 then
+      raise
+        (File_exceeds_max_size_limit
+           ("File exceeds max size limit: " ^ Int64.to_string max_size)
+        )
+  in
+  let file_name = head.Tar.Header.file_name in
+  debug "Unpacking tar file: %s" file_name ;
+  assert_file_path_is_legal file_name ;
+  let file_path = dir ^ file_name in
+  match head.Tar.Header.link_indicator with
+  | Tar.Header.Link.Normal ->
+      let file_size = head.Tar.Header.file_size in
+      let new_acc_size = acc_size + file_size in
+      Unixext.with_file file_path [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC]
+        file_perm (fun ofd ->
+          assert_file_not_exceed_max_size new_acc_size ;
+          Unixext.copy_file ?limit:(Some file_size) ifd ofd |> ignore
+      ) ;
+      new_acc_size
+  | Tar.Header.Link.Directory ->
+      let new_acc_size = acc_size + dir_size in
+      assert_file_not_exceed_max_size new_acc_size ;
+      Unixext.mkdir_rec file_path dir_perm ;
+      new_acc_size
+  | Hard
+  | Symbolic
+  | Character
+  | Block
+  | FIFO
+  | GlobalExtendedHeader
```

Comment:
The extended headers might be created if you create `tar` in PAX format. Might be useful looking into how to support them, unless we only want to unpack TAR files created by ourselves.

Reply:
The TAR files are created by ourselves. Confirmed with the operation team, in the current stage, we only support 'original' format TAR files (it means we do not support huge files, long file path, etc). So I think in the current stage, we can ignore these extended header.